### DISCUSSION TITLE
What is PyTest's order of execution in plugins?

### DISCUSSION THREAD

**Discussion by AbdouMechraoui** [pytest-dev/pytest#10532] [Q&A]
I would like to know the order of execution when writing a pytest plugin. In the [docs][1], this is done by using `tryfirst= True` or `trylast=True` for hooks. I understnd that plugins are nothing but a combination of hooks and fixtures. I would like to get deeper understanding of the flow of execution, as it doesn't seem sequential when running the follwoing code.
p.s. apologies if already asked and answered, would love a link :)
 My questions are as follow:

 - **What happens in case of multiple `tryfirst=True`?**
 - **when does the plugin script run sequentially?**
 - **what's the order of execution for the following code?**

```
# My Custom Plugin

def pytest_addoption(parser):
  # Register argparse-style options
    print("Addoption!")


def pytest_configure(config):
  # Allow plugins to perform initial configuration
    print("Configure!")


@pytest.hookimpl(tryfirst= True)
def pytest_report_header(config):
    """ First tryFirst! """
    print("First tryFirst!")


@pytest.hookimpl(trylast=True)
def pytest_terminal_summary(terminalreporter, exitstatus=6969):
    # will execute as late as possible
    print("TryLast!")
    return None

@pytest.hookimpl(tryfirst= True)
def pytest_report_header(config):
    """ Second tryFirst! """
    print("Second tryFirst!")


@pytest.hookimpl(hookwrapper=True)
def pytest_collection_modifyitems(items):
    # will execute even before the tryfirst one above!
   print("Wrapper before yield!")
    outcome = yield
    print("Wrapper after yield!")
    # will execute after all non-hookwrappers executed


```

  [1]: https://docs.pytest.org/en/7.1.x/how-to/writing_hook_functions.html#hook-function-ordering-call-example

**Comment by RonnyPfannschmidt** [comment]
We ought to link pluggy docs there

**Reply by RonnyPfannschmidt** [reply]
To quickly elaborate the order is random but the topology is grouped by 2 details

1. Hookwappers vs normal
2. Try first followed by unmarked followed by try last

The rest of the order is roughly defined by registration order

First come pytest core plugins, then entrypoints in discovery order , then conftests
